### PR TITLE
[Feature/CON-129] add callback function prop to accordionProduct in gel for tracking

### PR DESCRIPTION
**Pull request by hassanraza0054**
This PR addresses the tracking requirement in [CON-129](https://aesoponline.atlassian.net/browse/CON-129).

Work on gel involved:-

1. Passing product ref to accordionProduct
2. Enriching with forwardRef function
3. **Introducing new  _callbackWithRef_  function prop** that would allow any external function to have access to the ref of any gel component, generated within the gel components context
4. Improved AccordionProduct.js test coverage to 100% [function testing coverage remains 85% due to default prop. Will address in future PR]

Corresponding **web-ui** PR: https://github.com/aesop/aesop-web-ui/pull/3649

### CODE REVIEW

File: src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js [hunk 1]
```diff
@@ -115,6 +115,7 @@ const HorizontalProductDisplayAccordion = ({ id, products, addToCartCopy }) => {
                   resetAccordion={resetAccordion}
                   toggleAccordion={toggleAccordion}
                   {...product}
+                  ref={product.ref}
```
Comment (MkMan) [inline src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js:118]: Is this needed considering the `product` is spread on the above line?

File: src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js [hunk 2]
```diff
@@ -136,6 +137,7 @@ HorizontalProductDisplayAccordion.propTypes = {
       title: PropTypes.string,
     }),
   }),
+  onPromoClick: PropTypes.func,
```
Comment (MkMan) [inline src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js:140]: Maybe it's my new glasses but I can't usage of this prop in this component. Am I missing something?

File: src/components/HorizontalProductDisplayAccordion/components/AccordionProduct/AccordionProduct.fixture.js [hunk 3]
```diff
@@ -37,6 +37,52 @@ const AccordionProductFixture = {
...
+  closedStateWithVideo: {
+    addToCart: {
+      cartAction: 'Add to your cart',
+      updateNotification: 'added to your cart',
+      outOfStock: {
+        label: 'Out of Stock',
+        title: 'Out of stock',
+      },
+    },
+    background: 'Video',
```
Comment (MkMan) [inline src/components/HorizontalProductDisplayAccordion/components/AccordionProduct/AccordionProduct.fixture.js:49]: @beau-charman Do you think it's worthwhile to eventually have "shared" fixtures somewhere for the common components? For example, `Video`, `Image`, `Link` are used in many other components and often we copy/paste the fixture.

File: src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js (hunk 2 above)
Comment (hassanraza0054) [reply src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js:140]: Ah I missed removing that. Good catch! New glasses are A-ok haha

File: src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js (hunk 1 above)
Comment (zack76) [reply src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js:118]: > Is this needed considering the `product` is spread on the above line?

I guess `{...product}` includes the ref prop already

Comment (hassanraza0054) [reply src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js:118]: Yup good catch. Was unsure whether the spread operator automatically assigned it to the ref in the accordionProduct component given the way the props were destructured. Thanks for the input lads.

File: src/components/HorizontalProductDisplayAccordion/components/AccordionProduct/AccordionProduct.fixture.js (hunk 3 above)
Comment (hassanraza0054) [reply src/components/HorizontalProductDisplayAccordion/components/AccordionProduct/AccordionProduct.fixture.js:49]: Yup same question here!

File: src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js (hunk 1 above)
Comment (hassanraza0054) [reply src/components/HorizontalProductDisplayAccordion/HorizontalProductDisplayAccordion.js]: Changes made!

**Review by MkMan — APPROVED** [review]

**Pull request merged**